Edge tiling

spacewaffle opened the pull request:
This allows the inference server to take a 'tile' query parameter that looks like '300' in order to split up large images into smaller ones before it does inference. Steps are as follows:

1. User sends a POST request with `&tile=300` for 300x300 tiles.
2. If the tiling does not perfectly match image dimensions, we add some padding to the full image
3. Cut up the image into tiles (smaller tensors)
4. Process each image individually with detect()
5. Once we get predictions, modify the x,y coordinates of the predictions based on the tile position

Testing curl request for Stratos: (note: you might need to bump up their device limit in admin to test)

`base64 test4.jpg | curl -d @- "http://localhost:9001/test-2-5ukex/6?api_key=MH84j2Z99SNRktLcyS8q&tile=325"`

To verify tiled predictions are correct, output predictions to a json file with the same name as the image, then upload both to any project upload page on app.roboflow.com:

`base64 test.jpg | curl -d @- "http://localhost:9001/test-2-5ukex/6?api_key=MH84j2Z99SNRktLcyS8q&tile=325" > test.json`

Large image detecting cars example without tiling:
![Screen Shot 2022-06-13 at 10 43 58 AM](https://user-images.githubusercontent.com/1930218/173403725-d266be53-be5e-4c39-a51f-25d786414dcd.png)

Same image with 300px tiling
![Screen Shot 2022-06-13 at 10 44 45 AM](https://user-images.githubusercontent.com/1930218/173403789-96633df0-94e5-44c4-aaae-962c90118e77.png)

spacewaffle committed:
setup for new 'tile' query parameter

spacewaffle committed:
split image method

spacewaffle committed:
added pad image function

spacewaffle committed:
padding function

spacewaffle committed:
update splitImage function name

spacewaffle committed:
fix parameter references for splitImage

spacewaffle committed:
abstract out detect function and add tiling offsets for x and y

spacewaffle committed:
fix typo

spacewaffle committed:
replace abstracted function with tile check and loop through tiles to do inference

spacewaffle committed:
mutable variables

spacewaffle committed:
fixed up node wiring

spacewaffle committed:
make it possible to upload large images as lfs if needed

spacewaffle committed:
added some images

spacewaffle committed:
necessary comment

spacewaffle committed:
clean up tensor from memory after splitting

spacewaffle committed:
dispose of more tensors after usage

spacewaffle committed:
added tidy for infer()

spacewaffle committed:
fix console logging

spacewaffle committed:
for memory leak debugging

spacewaffle committed:
fixed issue with offset coordinates

spacewaffle committed:
test images

spacewaffle committed:
force tiles to be square and take a single number

spacewaffle committed:
removed all unnecessary files

spacewaffle committed:
clean up

spacewaffle committed:
add back correct start time to get proper infer timing

yeldarby commented on .gitattributes:
```diff
@@ -0,0 +1 @@
+*.jpg filter=lfs diff=lfs merge=lfs -text
```
Wasn't familiar with this. Checking understanding: This uses git large file system to store images?

Any additional dependencies we need or is this built into git already?

spacewaffle replied:
I might just take this out because this was meant for uploading the large images I was working with at the time. Since we're merging with this, I don't think we need git lfs anymore here.

yeldarby replied:
We prob want to gitignore any test images anyway; we don't want them in this public repo or getting wrapped up in our docker.

spacewaffle replied:
Yup true. Will add.

yeldarby commented on server/index.js:
```diff
@@ -56,74 +75,49 @@ app.get("/", function(req, res) {
 		req.model.totalTime = (req.model.totalTime||0)+(Date.now()-start);
 
+            var allowed_classes = null; // allow all
+        if(req.query.classes) {
```
Looks like spacing got messed up here.

spacewaffle replied:
This is showing in my editor as lined up, but not on github, and I am fully pushed. I'm not sure what's going on there...
![Screen Shot 2022-06-14 at 4 36 53 PM](https://user-images.githubusercontent.com/1930218/173700725-b44fadaa-52e2-4188-b7aa-18f6e5790557.png)

yeldarby commented on server/index.js:
```diff
@@ -56,74 +75,49 @@ app.get("/", function(req, res) {
                 return {
-                    x: Math.round(p.bbox.x * 10)/10,
-                    y: Math.round(p.bbox.y * 10)/10,
+                    x: x,
```
I think we still want to round to the nearest 1/10 of a pixel so floating point math doesn't blow up the response size with a bunch of insignificant precision & to keep the logic the same for the non-tiling scenario.

spacewaffle replied:
We still do round to the nearest 10th up on line 104.

yeldarby commented on server/index.js:
```diff
@@ -135,14 +129,121 @@ var infer = function(req, res) {
+
+        //takes a tile query parameter like so tile=3x5
+        if(req.query.tile) {
+            if(false){
```
Can remove stub?

yeldarby commented on server/index.js:
```diff
@@ -135,14 +129,121 @@ var infer = function(req, res) {
+            configuration.threshold = req.query.confidence;
+        }
+
+        //takes a tile query parameter like so tile=3x5
```
I think we changed this to pixels vs rows x cols

yeldarby commented on server/index.js:
```diff
@@ -135,14 +129,121 @@ var infer = function(req, res) {
+            //ensure the tile query param looks like tile=300
+            if (!req.query.tile.match(/([0-9]+)/)){
+                return res.json({
+                    error: "Tile query parameter improperly formatted."
```
This error message could be more helpful; we should tell the user what the format should be.

Would it be hard to take non-square tiles if the user wanted to maintain aspect ratio? (Eg if they trained their model on non-square tiles which got squished in the resize step to square we'd want that same thing to happen during inference for most accurate results.)

Recommendation would be also accepting eg `tile=300x200` format (and if they don't have the `x###` part we treat it as a square and use the first value for both dimensions.

spacewaffle replied:
Got the errors handled better.

Possible but don't know if that affects our performance. Will punt to a v2.

Also punting rectangular tiling to a v2 since the current tensor logic doesn't slice up tiles correctly when given rectangular dimensions.

yeldarby commented on server/index.js:
```diff
@@ -135,14 +129,121 @@ var infer = function(req, res) {
+            const paddedImage = padImage(tensor, tileDimensions);
+            const tiles = splitImage(paddedImage, tileDimensions);
+            var combinedResult = [];
+            async.eachOf(roboflow.tf.unstack(tiles), function(tile, index, cb) {
```
Does `eachOf` vs `eachOfSeries` or `eachOfLimit` produce better performance? Or are they all roughly equivalent? I'm wondering if there's a risk that spamming the event loop (and then the gpu) with several hundred callbacks in parallel might cause a bottleneck (or potentially make the GPU run out of memory) but maybe TFjs is handling that orchestration in the backend for us & this lets the GPU and CPU bound operations run independently.

spacewaffle replied:
Performance is roughly equivalent which was initially weird to me. I would have assumed firing all of them off at once would run the inferences in parallel but it doesn't seem to. I honestly don't think it's hitting the gpu any harder than normal because of this.

yeldarby commented on server/index.js:
```diff
@@ -135,14 +129,121 @@ var infer = function(req, res) {
+        } else {
+            detect(req, res, start, tensor, function(error, result){
+                if(error){
+                    console.log("error on detect", error)
```
Was this error previously flowed through to the response?

spacewaffle replied:
Yes it was, fixed in the latest version

yeldarby commented on server/index.js:
```diff
@@ -135,14 +129,121 @@ var infer = function(req, res) {
+		});
+	}
+
+    roboflow.tf.tidy(() => {
```
Spacing also messed up here.

yeldarby reviewed (approved):
Looks mostly good; I left some comments below but am good with merging as-is if we want to follow up on those at a later time. I leave it to your discretion.

spacewaffle committed:
remove git lfs stuff

spacewaffle committed:
style changes for pr

spacewaffle committed:
better error handling logic

spacewaffle commented:
Fixed the style issues, added better error handling logic, and simplified some of it with promises. If an individual tile has an inference error, it'll show up in the predictions array as a single error object. If there's some kind of larger error, we'll return an error message instead of the predictions array. Going to merge now.

spacewaffle committed:
these disposes are not needed

spacewaffle committed:
ignore common image files
